File: backend/app/schemas/metrics.py
"""Server metrics API schemas."""

from pydantic import BaseModel


class DiskUsage(BaseModel):
    mount: str
    size_bytes: int
    used_bytes: int
    pct: float


class DeviceMetricsSummary(BaseModel):
    """列表页:单台服务器最新指标摘要。"""

    device_id: int
    device_name: str
    ip_address: str | None = None
    type: str
    os_system: str | None = None
    rack_id: int | None = None
    rack_name: str = ""
    room_id: int | None = None
    room_name: str = ""
    status: str = "offline"  # monitor 的存活状态
    available: bool = False  # 本轮指标是否采集成功
    error: str | None = None
    fetched_at: str | None = None
    source: str | None = None  # ssh / winrm
    cpu_pct: float | None = None
    mem_pct: float | None = None
    disk_max_pct: float | None = None
    load1: float | None = None
    uptime_sec: int | None = None
    net_rx_bps: float | None = None
    net_tx_bps: float | None = None
    disk_read_bps: float | None = None
    disk_write_bps: float | None = None


class DeviceMetricsListResponse(BaseModel):
    items: list[DeviceMetricsSummary]
    total: int


class DeviceMetricsDetail(DeviceMetricsSummary):
    """详情抽屉:最新指标全量 + 分区明细。"""

    mem_used_mb: int | None = None
    mem_total_mb: int | None = None
    load5: float | None = None
    load15: float | None = None
    disks: list[DiskUsage] = []


class MetricSamplePoint(BaseModel):
    ts: str
    cpu_pct: float | None = None
    mem_pct: float | None = None
    disk_max_pct: float | None = None
    load1: float | None = None
    net_rx_bps: float | None = None
    net_tx_bps: float | None = None
    disk_read_bps: float | None = None
    disk_write_bps: float | None = None


class MetricsHistoryResponse(BaseModel):
    device_id: int
    range: str
    points: list[MetricSamplePoint]


class MetricTrendPoint(BaseModel):
    """概览曲线点:只带 CPU / 内存，避免多设备批量返回时载荷过大。"""

    ts: str
    cpu_pct: float | None = None
    mem_pct: float | None = None


class DeviceMetricTrend(BaseModel):
    """单台服务器的 CPU / 内存历史曲线。"""

    device_id: int
    device_name: str
    points: list[MetricTrendPoint] = []


class MetricsTrendResponse(BaseModel):
    """全部纳管设备(server/cloud_server/host)的曲线集合，一次请求返回。"""

    range: str
    total: int
    series: list[DeviceMetricTrend] = []

File: frontend/src/types/metrics.ts
/** 服务器指标监控类型(对应后端 /api/metrics/*) */

export interface MetricsDeviceItem {
  device_id: number
  device_name: string
  ip_address: string | null
  type: string
  os_system: string | null
  rack_id: number | null
  rack_name: string
  room_id: number | null
  room_name: string
  status: string
  available: boolean
  error: string | null
  fetched_at: string | null
  source: string | null // ssh / winrm
  cpu_pct: number | null
  mem_pct: number | null
  disk_max_pct: number | null
  load1: number | null
  uptime_sec: number | null
  net_rx_bps: number | null
  net_tx_bps: number | null
  disk_read_bps: number | null
  disk_write_bps: number | null
}

export interface MetricsDeviceList {
  items: MetricsDeviceItem[]
  total: number
}

export interface DiskUsage {
  mount: string
  size_bytes: number
  used_bytes: number
  pct: number
}

export interface MetricsDeviceDetail extends MetricsDeviceItem {
  mem_used_mb: number | null
  mem_total_mb: number | null
  load5: number | null
  load15: number | null
  disks: DiskUsage[]
}

export type MetricsRange = '1h' | '6h' | '24h' | '7d'

export interface MetricSamplePoint {
  ts: string
  cpu_pct: number | null
  mem_pct: number | null
  disk_max_pct: number | null
  load1: number | null
  net_rx_bps: number | null
  net_tx_bps: number | null
  disk_read_bps: number | null
  disk_write_bps: number | null
}

export interface MetricsHistory {
  device_id: number
  range: string
  points: MetricSamplePoint[]
}

/** 首页概览:全部服务器的 CPU / 内存曲线 */
export interface MetricTrendPoint {
  ts: string
  cpu_pct: number | null
  mem_pct: number | null
}

export interface DeviceMetricTrend {
  device_id: number
  device_name: string
  points: MetricTrendPoint[]
}

export interface MetricsTrend {
  range: string
  total: number
  series: DeviceMetricTrend[]
}
